engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnStateBehavior.java

```diff
@@ -225,4 +225,16 @@ public void propagateTemporaryVariables(
         variablesState.getVariablesAsDocument(sourceContext.getElementInstanceKey());
     variablesState.setTemporaryVariables(targetContext.getElementInstanceKey(), variables);
   }
+
+  public void transferTemporaryVariables(
+      final BpmnElementContext sourceContext, final long targetElementInstanceKey) {
+
+    final var variables =
+        variablesState.getTemporaryVariables(sourceContext.getElementInstanceKey());
+
+    if (variables != null) {
+      variablesState.setTemporaryVariables(targetElementInstanceKey, variables);
+      variablesState.removeTemporaryVariables(sourceContext.getElementInstanceKey());
```

Comment:
Do we remove the variables from the event subprocess so the process that's inside it propagates its variables up when it's completed?

Reply:
Yes. The variables are removed from the event subprocess context because they're already merged when the start event is completed. This is also important to not use them again when the event subprocess is completed :)